### PR TITLE
add(scheduler/framework): implement smaller Pod update events

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Implement new cluster events UpdatePodScaleDown and UpdatePodLabel for Pods.

We aim to replace all Pod/Update events registered in in-tree plugins so that we can reduce the number of events subscribed by the scheduler, thus reducing the amount of memory consumed for in-flight events.

EDIT(2024/Jun/28): Actually even after this PR there are some remaining Pod/Update events, which are added in https://github.com/kubernetes/kubernetes/pull/122234. I'll deal with them as a follow up so that I don't make this PR even bigger from where has been reviewed so far.

#### Which issue(s) this PR fixes:

Part of #120622

#### Does this PR introduce a user-facing change?

```release-note
Implement new cluster events UpdatePodScaleDown and UpdatePodLabel for scheduler plugins.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/eventhandlers.go [hunk 1]
```diff
@@ -522,6 +528,26 @@ func addAllEventHandlers(
...
+		events = append(events, queue.PodLabelChange)
+	}
+
+	if podResourceRequestChanged(newPod, oldPod) {
+		events = append(events, queue.PodRequestChange)
+	}
+
+	if len(events) == 0 {
+		events = append(events, queue.AssignedPodUpdate)
+	}
```
Comment (sanposhiho) [inline pkg/scheduler/eventhandlers.go:545]: We have to tell other updates as `AssignedPodUpdate` because custom plugins may want to subscribe to other kinds of updates in Pods. 
So, this PR itself doesn't contribute to a reduction of the number of events registered in inFlightPods. But, after https://github.com/kubernetes/kubernetes/pull/122629, this PR does.

File: pkg/scheduler/internal/queue/events.go [hunk 2]
```diff
@@ -35,17 +35,18 @@ const (
...
 	NodeAdd = framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add, Label: "NodeAdd"}
-	// AssignedPodUpdate is the event when a pod is updated that causes pods with matching affinity
-	// terms to be more schedulable.
+	// AssignedPodUpdate is the event when a pod is updated.
 	AssignedPodUpdate = framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Update, Label: "AssignedPodUpdate"}
-	// AssignedPodDelete is the event when a pod is deleted that causes pods with matching affinity
-	// terms to be more schedulable.
+	// AssignedPodDelete is the event when a pod is deleted.
 	AssignedPodDelete = framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete, Label: "AssignedPodDelete"}
+	// PodRequestChange is the event when a pod's label is changed.
```
Comment (AxeZhan) [inline pkg/scheduler/internal/queue/events.go:46]:
```suggestion
	// PodRequestChange is the event when a pod's resource request is changed.
```

File: pkg/scheduler/eventhandlers_test.go [hunk 3]
```diff
@@ -606,3 +604,79 @@ func TestNodeSchedulingPropertiesChange(t *testing.T) {
...
+		{
+			name:   "only pod's resource request is updated",
+			oldPod: podWithBigRequest,
+			newPod: podWithSmallRequest,
+			want:   []framework.ClusterEvent{queue.PodRequestChange},
+		},
+		{
+			name:   "different part of pod is updated",
+			newPod: st.MakePod().Annotation("foo", "bar").Obj(),
+			oldPod: st.MakePod().Annotation("foo", "bar").Obj(),
```
Comment (nayihz) [inline pkg/scheduler/eventhandlers_test.go:670]: nit: "bar2"

File: pkg/scheduler/eventhandlers.go (hunk 1 above)
Comment (kerthcet) [reply pkg/scheduler/eventhandlers.go:545]: Will we miss some combined pod events here, like pod annotations & pod label updates, as well as pod status.
We should make sure it's compatible.

Comment (sanposhiho) [reply pkg/scheduler/eventhandlers.go:545]: I don't think a combined event would be a problem here.
Let's say one Pod update contains the update in annotations & label - the update will be interpreted as `PodLabelChange`, and this `PodLabelChange` will be matched with Pod/Update as well = QHint registered for Pod/Update will be called with PodLabelChange event.

**Comment by sanposhiho** [comment]
@kubernetes/sig-scheduling-approvers Can anyone take a look?

**Review by kerthcet — COMMENTED** [review]
Thanks, this is a significant improve as we have specified pod events now, can we have an integration test to cover this? Also cc @alculquicondor @Huang-Wei as we changed the node register events as well.

File: pkg/scheduler/framework/types.go [hunk 4]
```diff
@@ -46,19 +46,24 @@ type ActionType int64
 
 // Constants for ActionTypes.
 const (
-	Add    ActionType = 1 << iota // 1
-	Delete                        // 10
+	Add ActionType = 1 << iota
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go:49]: Do we need the comment actually?

[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/framework/types.go:49]: If this can make it more readable, yes.

[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go:49]: > If this can make it more readable

I meant, can this? I don't see any usecase where people have to know the exact number of each event.

**Comment by sanposhiho** [comment]
Updated based on the suggestion from @kerthcet

File: pkg/scheduler/eventhandlers.go [hunk 5]
```diff
@@ -527,6 +532,26 @@ func addAllEventHandlers(
 	return nil
 }
 
+// podSchedulingPropertiesChange interprets the update of a pod and returns corresponding UpdatePodXYZ event(s).
+func podSchedulingPropertiesChange(newPod *v1.Pod, oldPod *v1.Pod) []framework.ClusterEvent {
```
Comment (alculquicondor) [inline pkg/scheduler/eventhandlers.go:536]: Could we return a single event and `|` the action types?

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 6]
```diff
@@ -1061,12 +1061,12 @@ func isPodResourcesResizedDown(pod *v1.Pod) bool {
 
 // AssignedPodUpdated is called when a bound pod is updated. Change of labels
 // may make pending pods with matching affinity terms schedulable.
-func (p *PriorityQueue) AssignedPodUpdated(logger klog.Logger, oldPod, newPod *v1.Pod) {
+func (p *PriorityQueue) AssignedPodUpdated(logger klog.Logger, oldPod, newPod *v1.Pod, event framework.ClusterEvent) {
 	p.lock.Lock()
 	if isPodResourcesResizedDown(newPod) {
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue.go:1066]: The event already has the information about whether there was a change of requests

File: pkg/scheduler/eventhandlers.go [hunk 7]
```diff
@@ -527,6 +532,26 @@ func addAllEventHandlers(
...
+
+	if labelsChanged(newPod.ObjectMeta, oldPod.ObjectMeta) {
+		events = append(events, queue.PodLabelChange)
+	}
+
+	if podResourceRequestChanged(newPod, oldPod) {
+		events = append(events, queue.PodRequestChange)
+	}
+
+	if len(events) == 0 {
```
Comment (alculquicondor) [inline pkg/scheduler/eventhandlers.go:547]: Shouldn't we always consider AssignedPodUpdate? Otherwise we might break out-of-tree plugins.

File: pkg/scheduler/eventhandlers.go (hunk 5 above)
Comment (kerthcet) [reply pkg/scheduler/eventhandlers.go:536]: I have talked with @sanposhiho before, this will impact the `queue_incoming_pods_total` metrics since we don't have a `PodLableAndRequestChanged`(label name of the event) event.

Comment (alculquicondor) [reply pkg/scheduler/eventhandlers.go:536]: I think those events aren't that important. Maybe we can use the more generic PodUpdated

Comment (kerthcet) [reply pkg/scheduler/eventhandlers.go:536]: Have no idea whether will this harm users. Maybe we need feedbacks.

Comment (sanposhiho) [reply pkg/scheduler/eventhandlers.go:536]: Is such label value removal accepted in general? (If Yes, I'd rather follow Aldo's way.)
I agree those labels aren't very important, but at the same time, as @kerthcet said, it'd be a breaking change for those who use those labels.

File: pkg/scheduler/eventhandlers.go (hunk 7 above)
Comment (sanposhiho) [reply pkg/scheduler/eventhandlers.go:547]: Yes,

> the update will be interpreted as PodLabelChange, and this PodLabelChange will be matched with Pod/Update as well = QHint registered for Pod/Update will be called with PodLabelChange event. 
https://github.com/kubernetes/kubernetes/pull/122628#discussion_r1451646204.

Comment (sanposhiho) [reply pkg/scheduler/eventhandlers.go:547]: Updated.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 6 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: Updated.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 8]
```diff
@@ -1045,28 +1045,14 @@ func (p *PriorityQueue) AssignedPodAdded(logger klog.Logger, pod *v1.Pod) {
...
-	}
-	return false
-}
-
 // AssignedPodUpdated is called when a bound pod is updated. Change of labels
 // may make pending pods with matching affinity terms schedulable.
-func (p *PriorityQueue) AssignedPodUpdated(logger klog.Logger, oldPod, newPod *v1.Pod) {
+func (p *PriorityQueue) AssignedPodUpdated(logger klog.Logger, oldPod, newPod *v1.Pod, event framework.ClusterEvent) {
 	p.lock.Lock()
-	if isPodResourcesResizedDown(newPod) {
```
Comment (kerthcet) [inline pkg/scheduler/internal/queue/scheduling_queue.go:1066]: Plz add back the todo here because what we really want here is **size down** if I understand correctly.

**Review by kerthcet — COMMENTED** [review]
Generally LGTM but still one concern here as mentioned at https://github.com/kubernetes/kubernetes/pull/122629#issuecomment-1951117940

TL;DR: We now support small events in scheduler, but currently all the events will be registered in hint map, therefore, although we'll process with `isEventOfInterest` but the result is always true, to avoid this, I think we should add more smaller events, then it looks like:
```
switch event:
    - case: podRequestUpdate
    - case: podLabelUpdate
    - case: podStatusUpdate
    - case: podAnnotationUpdate
    - default: general podUpdate event
```
then we can still remain the general update event(for default option) and filter out unrelated ones mostly like podStatusUpdate, this can be a follow-up.

**Comment by sanposhiho** [comment]
> We now support small events in scheduler, but currently all the events will be registered in hint map, therefore, although we'll process with isEventOfInterest but the result is always true

No, after this PR, only smaller Pod events are registered in hint map and `isEventOfInterest` will start to filter a general Pod update events.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 8 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:1066]: Yes, but that's already handled by QHint of the resource fit plugin. So, nothing else to do, I think.

**Comment by kerthcet** [comment]
> No, after this PR, only smaller Pod events are registered in hint map and `isEventOfInterest` will start to filter out general Pod update events.

Unless we remove the `AssignedPodUpdate` in https://github.com/kubernetes/kubernetes/blob/d19bf7c7b79af1e84c10ad5caed4b0fcd9980b85/pkg/scheduler/eventhandlers.go#L536-L552

otherwise it's always matched, but we talked offline that we will not remove this.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 8 above)
Comment (kerthcet) [reply pkg/scheduler/internal/queue/scheduling_queue.go:1066]: Fair to me now because we didn't make the code worse, we didn't implement **resize down** logic before as well.

But in the long run, I think this is a matter of choice, we can always leave the judgement to the hint functions, e.g. replace below with `p.moveAllToActiveOrBackoffQueue` 

```
p.movePodsToActiveOrBackoffQueue(logger, p.getUnschedulablePodsWithMatchingAffinityTerm(logger, newPod), AssignedPodUpdate, oldPod, newPod)
```

**OR** we have another **simple** filter logic ahead of hint function like we will judge _whether resize downed here_ instead of _whether the request changed_ for the sake of a bit of performance, maybe huge in a large cluster.

Poor points.

**Comment by sanposhiho** [comment]
@kerthcet 

Sorry, I completely misunderstood the current implementation, you're right. 
I somehow thought that when a general Pod update event comes, it wouldn't trigger QHint registered with PodRequestChanges, while when PodRequestChanges comes, it trigger QHint registered with a general Pod update event.

But, actually, doesn't it make sense? If a plugin registers PodRequestChanges, they want to subscribe to PodRequestChanges specifically and aren't interested in other changes in Pod, which are interpreted to a general Pod update event in `podSchedulingPropertiesChange`. 
If a plugin registers a general Pod update event, they want to subscribe to all Pod related changes and we'll trigger them with all Pod update events, including a general one, PodRequestChanges, and PodLabelUpdate.

**Comment by sanposhiho** [comment]
I mean, like this: https://github.com/kubernetes/kubernetes/pull/122628/commits/94b430241c300ee04f819faa1041eb0c393fe289.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 8 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:1086]: I agree with the first point; we should remove all pre-filtering stuff (including your example and preCheck) and always leave the judgement to QHint.

**Comment by kerthcet** [comment]
Sorry for the delayed response, hope we can still catch the code freeze.

I wouldn't like to break the semantic of general pod update event which means any possible part of pod update. Based on the assumption, I think add a new `podStatusUpdate` event can already filter out most of the events. This is more intuitive to me. WDYT?

**Comment by sanposhiho** [comment]
@kerthcet 

> I wouldn't like to break the semantic of general pod update event 

It wouldn't break the meaning of a general pod update event from users' pov. 
What I'm proposing in short is:
- EventsToRegister has a general Pod update → all Pod events (including PodRequestChanges etc) trigger QHint.
- EventsToRegister has PodRequestChanges → only PodRequestChanges triggers QHint. It makes sense because we interpret Pod changes that aren't PodRequestChanges or podLabelUpdate into a general Pod update in `podSchedulingPropertiesChange`.

**Comment by sanposhiho** [comment]
I actually wouldn't like to define smaller Pod events other than PodRequestChanges and podLabelUpdate until we actually need them. Let's say here we define `podStatusUpdate`, we may want to break `podStatusUpdate` into further smaller ones in the future like `podStatusConditionUpdate` etc.

**Comment by kerthcet** [comment]
> Let's say here we define podStatusUpdate, we may want to break podStatusUpdate into further smaller ones in the future like podStatusConditionUpdate etc

Fair, maybe we can follow how *Node Event* did: `Update = UpdateNodeAllocatable | UpdateNodeLabel | UpdateNodeTaint | UpdateNodeCondition | UpdateNodeAnnotation` 

cc @alculquicondor agree with this approach https://github.com/kubernetes/kubernetes/commit/94b430241c300ee04f819faa1041eb0c393fe289.?

**Comment by alculquicondor** [comment]
We are too close to the code freeze, so I would like to focus on critical features and fixes.

**Comment by sanposhiho** [comment]
Just rebased.

**Comment by kerthcet** [comment]
/assign
I'll take a look tonight or tomorrow. This can notify me to finish the review work.

**Comment by sanposhiho** [comment]
(just rebased)

File: pkg/scheduler/framework/types.go [hunk 9]
```diff
@@ -164,14 +184,25 @@ func (ce ClusterEvent) IsWildCard() bool {
 }
 
 // Match returns true if ClusterEvent is matched with the coming event.
+// "match" means the coming event is the same or more specific than the ce.
+// i.e., when ce.ActionType is Update, it return true if a coming event is UpdateNodeLabel
```
Comment (kerthcet) [inline pkg/scheduler/framework/types.go:188]: OK, so this is the last part we may argue about, we have the power of interpretation but have no idea whether this will break users. cc @alculquicondor @Huang-Wei who have contextual knowledge.

Comment (kerthcet) [reply pkg/scheduler/framework/types.go:188]: Maybe not matter because we already categoties in the front when event comes, like the node events.

Comment (kerthcet) [reply pkg/scheduler/framework/types.go:188]: For a combinatorial event, e.g. podLabel & podRequest， it should match with each event I think.

File: pkg/scheduler/framework/plugins/noderesources/fit_test.go [hunk 10]
```diff
@@ -1095,7 +1095,7 @@ func TestEventsToRegister(t *testing.T) {
 			"Register events with InPlacePodVerticalScaling feature enabled",
 			true,
 			[]framework.ClusterEventWithHint{
-				{Event: framework.ClusterEvent{Resource: "Pod", ActionType: framework.Update | framework.Delete}},
+				{Event: framework.ClusterEvent{Resource: "Pod", ActionType: framework.UpdatePodRequest | framework.Delete}},
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/noderesources/fit_test.go:1098]: That is a good point. So, you mean we can create `PodRequestScaledDown` instead of `UpdatePodRequest`.

The difference would be:
- `PodRequestScaledDown`: Filter out events that represent scaling up in `podSchedulingPropertiesChange`.
- `UpdatePodRequest`: Filter out such events in QHint of resource fit plugin instead.

Looks similar, but `PodRequestScaledDown` has a certain advantage actually; it filters out events before putting events in inFlightEvents. So, it could help reducing the memory consumed by inFilghtEvents for sure. (https://github.com/kubernetes/kubernetes/issues/120622)

@alculquicondor What do you think; `UpdatePodRequest` vs `PodRequestScaledDown`. Any particular concern for `PodRequestScaledDown`?

File: pkg/scheduler/framework/types.go [hunk 11]
```diff
@@ -82,6 +96,11 @@ const (
 	// unschedulable pod pool.
 	// This behavior will be removed when we remove the preCheck feature.
 	// See: https://github.com/kubernetes/kubernetes/issues/110175
+	//
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go:99]: Yeah, I moved the comment to L77, along with applying your suggestion https://github.com/kubernetes/kubernetes/pull/122628#discussion_r1579298886.

**Comment by sanposhiho** [comment]
Had to rebased to fix the issue in testing.

File: pkg/scheduler/framework/types.go (hunk 9 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go:183]: It shouldn't be a concern; Let's have an example with Pod/Update and Pod/PodLabelUpdate:
- if a user registers Pod/Update, it matches if an incoming event is Pod/PodLabelUpdate (actually any types of Pod update events).
- if a user registers Pod/PodLabelUpdate, it matches **_only_** if an incoming event is Pod/PodLabelUpdate. It doesn't match if an incoming event is Pod/Update.
  - Before this PR, it matches in the latter case too, which is unnecessary match.
    - Why unnecessary match? > because the scheduler interprets an event as Pod/PodLabelUpdate if a Pod got changed in labels. This means that if receiving a Pod/Update event, it's always an event that doesn't change the labels of the Pod.

**Comment by sanposhiho** [comment]
After https://github.com/kubernetes/kubernetes/pull/122234, we should create `UpdatePodSchedulingGates`.

**Comment by sanposhiho** [comment]
@kubernetes/sig-scheduling-approvers can anyone take a look at this PR too? 🙏

File: pkg/scheduler/internal/queue/events.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/events.go:1]: Most changes in this file is moved from eventhandlers.go. I had to do so to use `PodSchedulingPropertiesChange` in scheduling_queue.go.

Only the function `PodSchedulingPropertiesChange` (and its dependency) is the new functions introduced in this PR.

File: pkg/scheduler/internal/queue/events_test.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/events_test.go:1]: Ditto, most changes are just moved from eventhandlers.go.
Only the test `Test_podSchedulingPropertiesChange` is a new one.

**Comment by sanposhiho** [comment]
I'll do https://github.com/kubernetes/kubernetes/pull/122628#issuecomment-2132103320 as a follow up, so that I don't make this PR even bigger from where has been reviewed so far.

**Comment by sanposhiho** [comment]
cc @gabesaba 
(wondering if you could have time reviewing this too)

**Comment by sanposhiho** [comment]
Just rebased.

**Comment by sanposhiho** [comment]
@kubernetes/sig-scheduling-approvers @gabesaba 
Can anyone take a look? 🙏

File: pkg/scheduler/eventhandlers_test.go
Comment (gabesaba) [inline pkg/scheduler/eventhandlers_test.go:1]: are we losing any coverage, or were these tests moved to `pkg/scheduler/internal/queue/events_test.go`?

File: pkg/scheduler/internal/queue/events.go [hunk 12]
```diff
@@ -89,3 +96,85 @@ var (
...
+	}
+	if nodeTaintsChanged(newNode, oldNode) {
+		events = append(events, NodeTaintChange)
+	}
+	if nodeConditionsChanged(newNode, oldNode) {
+		events = append(events, NodeConditionChange)
+	}
+	if nodeAnnotationsChanged(newNode, oldNode) {
+		events = append(events, NodeAnnotationChange)
+	}
```
Comment (gabesaba) [inline pkg/scheduler/internal/queue/events.go:141]: could we do something like this (obviously with better names)
```go
changedFns := [nodeSpecUnschedulableChanged, nodeAllocatableChanged, ...]
for _, fn := range changedFns {
  event := fn(newNode, oldNode)
  if event != nil {
    events = append(events, event)
  }
}
```

 I think this increases readability, reduces repetition, and reduces risk that changed function returns the wrong event.

Also `PodSchedulingPropertyChanges`, can share this code - some common function - `func(newPod, oldPod, []Fns) []framework.ClusterEvent`

**Comment by sanposhiho** [comment]
also cc @macsko if you have time.. I know I've asked you many... 😅

File: pkg/scheduler/internal/queue/events.go
Comment (gabesaba) [reply pkg/scheduler/internal/queue/events.go:1]: Consider putting this move in a separate commit (or PR). Then, reviewers could focus on the relevant logic changes, rather than sorting through the noise for where the actual change is

**Review by gabesaba — COMMENTED** [review]
A few comments to start with. I will spend more time understanding the change before I do a more thorough review.

The only real logic change is in `scheduling_queue.go`, right? the rest of it is moving code (or adding tests)?

File: pkg/scheduler/eventhandlers_test.go
Comment (sanposhiho) [reply pkg/scheduler/eventhandlers_test.go:1]: Yes, they're just moved.

> Ditto, most changes are just moved from eventhandlers.go.
Only the test Test_podSchedulingPropertiesChange is a new one.
https://github.com/kubernetes/kubernetes/pull/122628#discussion_r1638920876

File: pkg/scheduler/internal/queue/events.go
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/events.go:1]: Created #126050

File: pkg/scheduler/internal/queue/events.go (hunk 12 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/events.go:141]: Put the refactoring together in #126050.

**Comment by sanposhiho** [comment]
/hold

Now this PR depends on https://github.com/kubernetes/kubernetes/pull/126050

File: pkg/scheduler/scheduler_test.go [hunk 13]
```diff
@@ -848,10 +849,10 @@ func Test_UnionedGVKs(t *testing.T) {
 			want: map[framework.GVK]framework.ActionType{},
 		},
 		{
-			name:    "plugins with default profile",
+			name:    "plugins with default profile (InPlacePodVerticalScaling: disabled)",
 			plugins: schedulerapi.PluginSet{Enabled: defaults.PluginsV1.MultiPoint.Enabled},
 			want: map[framework.GVK]framework.ActionType{
-				framework.Pod:                   framework.All,
+				framework.Pod:                   framework.Add | framework.UpdatePodLabel | framework.Delete,
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/scheduler_test.go:864]: Oh, this could potentially be very significant, as Pods have lots of status updates.
We wouldn't be able to observe this on the integration benchmarks (unless we add a case with status updates).

But it might increase the throughput in the load tests https://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=Scheduler&metricname=LoadSchedulingThroughput&TestName=load

@macsko FYI, for when you run our internal scale tests too.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 8 above)
Comment (alculquicondor) [reply pkg/scheduler/internal/queue/scheduling_queue.go:1176]: what if QHints are disabled? Is there anything else filtering for size down specifically?

File: pkg/scheduler/framework/types.go [hunk 14]
```diff
@@ -46,25 +46,39 @@ type ActionType int64
...
 
 // GVK is short for group/version/kind, which can uniquely represent a particular API resource.
 type GVK string
 
 // Constants for GVKs.
+//
+// Note:
+// - UpdatePodXYZ or UpdateNodeXYZ: triggered by updating particular parts of a Pod or a Node, e.g. updatePodLabel.
+// Use specific events rather than general ones (updatePodLabel vs update) can make the requeueing process more efficient
+// and consume less memories as less events will be cached at scheduler.
```
Comment (alculquicondor) [inline pkg/scheduler/framework/types.go:81]:
```suggestion
// and consume less memory as less events will be cached at scheduler.
```

File: pkg/scheduler/framework/types.go (hunk 9 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/types.go:194]: uhm.... maybe we need `UpdatePodOther` that gets added when no other `podChangeExtracters` find a change?

`comingEvent.ActionType <= ce.ActionType` just looks very unintuitive

File: pkg/scheduler/internal/queue/events.go [hunk 15]
```diff
@@ -89,6 +96,44 @@ var (
...
+		extractPodResourceRequestChange,
+	}
+
+	for _, fn := range podChangeExtracters {
+		if event := fn(newPod, oldPod); event != nil {
+			events = append(events, *event)
+		}
+	}
+
+	if len(events) == 0 {
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/events.go:113]: No, it's not a breaking change because Pod smaller events `Match()` with Pod/Update event, meaning, when plugins have Pod/Update event, they will still receive all events of Pod/update, including smaller events.

File: pkg/scheduler/scheduler_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/scheduler_test.go:864]: Exactly, it's the core motivation that we reduce the number of events in-tree plugins subscribing and thus reduce the amount of events that the scheduling queue has to store as `inFlightEvents`.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 8 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:1176]: > Is there anything else filtering for size down specifically?

In case QHint is disabled, No.

But, this topic is closely related to https://github.com/kubernetes/kubernetes/pull/122628#discussion_r1593835035
> @alculquicondor What do you think; UpdatePodRequest vs PodRequestScaledDown. Any particular concern for PodRequestScaledDown?

Maybe we can define `PodRequestScaledDown` instead of `UpdatePodRequest`. What do you think? @alculquicondor

File: pkg/scheduler/eventhandlers.go [hunk 16]
```diff
@@ -225,7 +227,10 @@ func (sched *Scheduler) updatePodInCache(oldObj, newObj interface{}) {
 		logger.Error(err, "Scheduler cache UpdatePod failed", "pod", klog.KObj(oldPod))
 	}
 
-	sched.SchedulingQueue.AssignedPodUpdated(logger, oldPod, newPod)
+	events := podSchedulingPropertiesChange(newPod, oldPod)
+	for _, evt := range events {
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/eventhandlers.go:242]: > Although in theory, it shouldn't happen more than once, as it is unlikely that a single update would change both requests and labels.

I don't concern very much because of this reason.

But, eventually it's better to aggregate all changes in the update and send only one aggregated event for SchedulingQueue. However it's not that simple to do so right away because some functions in SchedulingQueue use event.Label for metric label and logging. So, I'd like to keep here as it is, and do them in another PR as a follow up.

File: pkg/scheduler/framework/types.go (hunk 9 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go]: Implemented `UpdatePodOther` with a caution to users.

**Comment by sanposhiho** [comment]
@alculquicondor Updated with your suggestions and replied to your comments.

File: pkg/scheduler/framework/types.go [hunk 17]
```diff
@@ -46,25 +46,50 @@ type ActionType int64
...
+	// for better performance in requeueing.
+	UpdatePodLabel
+	// UpdatePodRequest is a update for pod's resource request calculated by resource.PodRequests() function.
+	UpdatePodRequest
+	// UpdatePodOther is a update for pod's other fields.
+	// NOT RECOMMENDED using it in your plugin's EventsToRegister,
+	// use Pod/Update instead when you have to subscribe Pod updates which are not covered by other UpdatePodXYZ events.
+	// Otherwise, your plugin might be broken when the upstream supports a new Pod specific Update event.
+	// It's used only for the internal event handling.
+	UpdatePodOther
```
Comment (alculquicondor) [inline pkg/scheduler/framework/types.go:76]: Maybe it's worth keeping it unexported... Then users can only use `Update`. But it still fullfills the purpose of making the matching logic more intuitive.

**Review by alculquicondor — COMMENTED** [review]
/approve
/hold for nit

**Comment by alculquicondor** [comment]
/lgtm
(in case you disagree with my nit)

File: pkg/scheduler/framework/types.go (hunk 17 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go]: Agree. Done 6fbe7714357b872c3b65b579e9eda7eb8f603d1c, but it requires another big diff to move pkg/scheduler/internal/queue/events.go under the framework package. 
(let me know if you prefer to do it in another follow-up PR, rather making this PR bigger.)

File: pkg/scheduler/framework/events.go
Comment (sanposhiho) [inline pkg/scheduler/framework/events.go:1]: have to move it under framework package to use unexported `updatePodOther`.
https://github.com/kubernetes/kubernetes/pull/122628#pullrequestreview-2186417175

File: pkg/scheduler/framework/types.go (hunk 17 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go]: (I had to force-push to solve a conflict made by the last commit.)

**Comment by sanposhiho** [comment]
@alculquicondor Also, can you revisit the discussion at https://github.com/kubernetes/kubernetes/pull/122628/files#r1682661244?

File: pkg/scheduler/framework/plugins/noderesources/fit_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/noderesources/fit_test.go:1098]: Yeah, I prefer we have PodRequestScaledDown, as scale ups won't make anything schedulable.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 8 above)
Comment (alculquicondor) [reply pkg/scheduler/internal/queue/scheduling_queue.go:1184]: Yes, let's do that.

Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:1179]: Done.

**Comment by sanposhiho** [comment]
Rebased + supported UpdatePodScaleDown, which replaces UpdatePodRequest.

**Comment by sanposhiho** [comment]
/unhold

Merge this PR before the code freeze.
Discussed with @macsko; Aldo already approved, except for the last few commits, and those last parts are approved by @macsko.
